### PR TITLE
Autoskip

### PR DESCRIPTION
**Pull request by xandkar**

### CODE REVIEW

File: src/handlers/miner_hbbft_handler.erl [hunk 1]
```diff
@@ -15,28 +15,45 @@
         ]).
 
 -include_lib("blockchain/include/blockchain_vars.hrl").
+-include_lib("helium_proto/include/blockchain_block_pb.hrl").
```
[the remark this answers is not in the document]
Comment (xandkar) [reply src/handlers/miner_hbbft_handler.erl:18]: It is used in `state_remove_var_txns_from_artifact/1` function to access `#blockchain_block_v1_pb{}` record.

File: src/handlers/miner_hbbft_handler.erl [hunk 2]
```diff
@@ -15,28 +15,45 @@
...
 -include("miner_util.hrl").
 
+-define(else, true).
```
[the remark this answers is not in the document]
Comment (xandkar) [reply src/handlers/miner_hbbft_handler.erl:21]: OK. Such a rare treat to stumble on a legit use case for `if` though :joy:

File: src/miner.erl [hunk 3]
```diff
@@ -215,6 +216,16 @@ relcast_queue(Group) ->
             end
     end.
 
+-spec sign(Artifact :: binary()) ->
```
[the remark this answers is not in the document]
Comment (xandkar) [reply src/miner.erl:219]: Yeah, this wasn't the wisest choice - will update.

File: test/miner_blockchain_SUITE.erl [hunk 4]
```diff
@@ -112,7 +126,7 @@ init_per_testcase(TestCase, Config0) ->
     DKGResults = miner_ct_utils:initial_dkg(Miners, Txns, Addresses, NumConsensusMembers, Curve),
     true = lists:all(fun(Res) -> Res == ok end, DKGResults),
 
-    timer:sleep(500),
+    timer:sleep(3000),
```
[the remark this answers is not in the document]
Comment (xandkar) [reply test/miner_blockchain_SUITE.erl:129]: I very much agree in principle and hated doing this. This setup code is not the most straight-forward to work with though, so I just followed the precedent for the time being. Will see if I can integrate Andrew's update.

File: test/miner_blockchain_SUITE.erl [hunk 5]
```diff
@@ -141,7 +155,82 @@ init_per_testcase(TestCase, Config0) ->
...
+        Txn = blockchain_txn_vars_v1:proof(Txn0, Proof),
+        miner_ct_utils:submit_txn(Txn, MinersInConsensus)
+    end)(),
+
+    AllHeights =
+        fun() ->
+            lists:usort([H || {_, H} <- miner_ct_utils:heights(MinersAll)])
+        end,
 
+    timer:sleep(10000),
```
[the remark this answers is not in the document]
Comment (xandkar) [reply test/miner_blockchain_SUITE.erl:213]: In retrospect - this isn't actually needed here.

File: test/miner_blockchain_SUITE.erl [hunk 6]
```diff
@@ -141,7 +155,82 @@ init_per_testcase(TestCase, Config0) ->
...
+            50,
+            1000
+        ),
+        "Heights equalized."
+    ),
+    [Height1] = AllHeights(),
+    ?assertMatch(
+        ok,
+        miner_ct_utils:wait_for_gte(height, MinersAll, Height1 + 1),
+        "Chain has advanced, so autoskip must've worked."
```
[the remark this answers is not in the document]
Comment (xandkar) [reply test/miner_blockchain_SUITE.erl:231]: Makes perfect sense. Will do.

**Comment by xandkar** [comment]
> this is too much.
> 
> can you please separate the feature and the refactoring of the internals into two different PRs?

Yes, true - this is much more than essential, but it was necessary to see where the essential parts should go, since important state updates were sprinkled throughput. Will split.

**Comment by xandkar** [comment]
> > this is too much.
> > can you please separate the feature and the refactoring of the internals into two different PRs?
> 
> Yes, true - this is much more than essential, but it was necessary to see where the essential parts should go, since important state updates were sprinkled throughput. Will split.

When 799 and 815 are done - I'll integrate them into the refactor (omitting the var autoskip) and then recreate the var autoskip PR.

One of the nice things of this refactor is that var autoskip can be easily enabled/disabled. Not sure if the same will be doable for timed autoskip, but would be great for testing if yes.

**Comment by evanmcc** [comment]
time based can easily be disabled by setting the timer very high

**Comment by xandkar** [comment]
Superseded by https://github.com/helium/miner/pull/822 and https://github.com/helium/miner/pull/823

**Pull request closed without merging**